### PR TITLE
[ Product Expiry ] Converted into new api.

### PR DESCRIPTION
**Pull request by jir-odoo**
Task : https://www.odoo.com/web?db=openerp#id=12034&view_type=form&model=project.task&action=327
Pad : https://pad.odoo.com/p/openerp-project.task-25OPZ1WVFK

Thanks.

### CODE REVIEW

File: addons/product_expiry/models/product_expiry.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import datetime
+from openerp import api, fields, models 
+
+
+class StockProductionLot(models.Model):
+    _inherit = 'stock.production.lot'
+    
+    @api.model
+    def _get_date(self, dtype):
+        """Return a function to compute the limit date for this type"""
```
Comment (rga-odoo) [inline addons/product_expiry/models/product_expiry.py]: it's return datetime string not function

File: addons/product_expiry/models/product_expiry.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        def calc_date(dtype):
```
Comment (rga-odoo) [inline addons/product_expiry/models/product_expiry.py]: there is no need of inner def calc_date

File: addons/product_expiry/models/product_expiry.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            """Compute the limit date for a given date"""
+            if not self.env.context.get('product_id', False):
```
Comment (rga-odoo) [inline addons/product_expiry/models/product_expiry.py]: product_id = self.env.context.get('product_id')

File: addons/product_expiry/models/product_expiry.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                date = False
+            else:
+                product = self.env['product.product'].browse(self.env.context.get('product_id'))
+                duration = getattr(product, dtype)
+                # set date to False when no expiry time specified on the product
+                date = duration and (datetime.datetime.today()
+                    + datetime.timedelta(days=duration))
+            return date and date.strftime('%Y-%m-%d %H:%M:%S') or False
```
Comment (rga-odoo) [inline addons/product_expiry/models/product_expiry.py]: date.strftime('%Y-%m-%d %H:%M:%S') ==> fields.Datetime.to_string(date)

File: addons/product_expiry/views/product_expiry_view.xml [hunk 5]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" ?>
-<openerp>
+<odoo>
     <data>
         <record model="ir.ui.view" id="view_move_form_expiry">
             <field name="name">stock.production.lot.form</field>
```
Comment (rga-odoo) [inline addons/product_expiry/views/product_expiry_view.xml]: stock.production.lot.form.inherit

File: addons/product_expiry/models/product_expiry.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        help='This is the date on which the goods with this Serial Number should be removed from the stock.')
+    alert_date = fields.Datetime(default=lambda self: self._get_date('alert_time'),
+        help="This is the date on which an alert should be notified about the goods with this Serial Number.")
+
+    # Assign dates according to products data
+    @api.model
+    def create(self, vals):
+        lot_rec = super(StockProductionLot, self).create(vals)
+        towrite = []
+        for field in ('life_date', 'use_date', 'removal_date', 'alert_date'):
```
Comment (hmo-odoo) [inline addons/product_expiry/models/product_expiry.py]: can you replace with

towrite = [field for field in ('life_date', 'use_date', 'removal_date', 'alert_date') if not getattr(lot_rec, field)]

File: addons/product_expiry/models/product_expiry.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        help='This is the date on which the goods with this Serial Number start deteriorating, without being dangerous yet.')
+    removal_date = fields.Datetime(default=lambda self: self._get_date('removal_time'),
+        help='This is the date on which the goods with this Serial Number should be removed from the stock.')
+    alert_date = fields.Datetime(default=lambda self: self._get_date('alert_time'),
+        help="This is the date on which an alert should be notified about the goods with this Serial Number.")
+
+    # Assign dates according to products data
+    @api.model
+    def create(self, vals):
+        lot_rec = super(StockProductionLot, self).create(vals)
```
Comment (hmo-odoo) [inline addons/product_expiry/models/product_expiry.py]: do not use "*_rec" or "*_obj". 
replace with just 'production_lot'

**Comment by hmo-odoo** [comment]
rename data/product_expiry_demo.xml ==> data/product_product_demo.xml.

make python file: stock_production_lot.py for inherit model "stock.production.lot" and inherit model "stock.quant".
make python file: product_template.py for inherit model "product.template"

rename views/product_expiry_view.xml ==> views/product_template_view.xml for inherit view  "product.template" and also make new view file for inherit view "stock.production.lot"

File: addons/product_expiry/models/stock_production_lot.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        production_lot = super(StockProductionLot, self).create(vals)
+        towrite = [field for field in ('life_date', 'use_date', 'removal_date', 'alert_date') if not getattr(production_lot, field)]
+        production_lot.write(self.with_context({'product_id': production_lot.product_id.id}).default_get(towrite))
+        return production_lot
+
+
+class StockQuant(models.Model):
+    _inherit = 'stock.quant'
+
+    removal_date = fields.Datetime(related='lot_id.removal_date', store=True)
```
Comment (hmo-odoo) [inline addons/product_expiry/models/stock_production_lot.py]: removal_date should be depend on lot_id and removal_date. so need to make depend method

Comment (jir-odoo) [inline addons/product_expiry/models/stock_production_lot.py]: No need to define depend method on related fields, With new API the value of the related field
will be automatically updated based on related="OTHER FIELD" parameter on fields.

Comment (msh-odoo) [inline addons/product_expiry/models/stock_production_lot.py]: :+1:  https://github.com/odoo/enterprise/blob/master/openerp/fields.py#L236

**Pull request closed without merging**